File: packages/shared/src/schemas/colorContrast.ts
// Mirrors services/langgraph/agency/artifacts/color_contrast.py. Same formula,
// same thresholds -- if these two drift, BrandCoreSchema's distinguishability
// check would pass or fail differently in the browser than in the backend
// that actually persists the artifact.
//
// WCAG 2.2 SC 1.4.11 exempts logotypes from contrast requirements ("Text
// that is part of a logo or brand name has no minimum contrast requirement").
// These utilities are not wired into LogoLockupSchema for that reason.

export const HEX_COLOR_RE = /^#[0-9A-Fa-f]{6}$/;

function srgbChannelToLinear(channel255: number): number {
  const c = channel255 / 255;
  if (c <= 0.03928) {
    return c / 12.92;
  }
  return ((c + 0.055) / 1.055) ** 2.4;
}

export function relativeLuminance(hexColor: string): number {
  if (!HEX_COLOR_RE.test(hexColor)) {
    throw new Error(`'${hexColor}' is not a 6-digit hex color (e.g. '#1a2b4c')`);
  }
  const value = hexColor.slice(1);
  const r = parseInt(value.slice(0, 2), 16);
  const g = parseInt(value.slice(2, 4), 16);
  const b = parseInt(value.slice(4, 6), 16);
  return (
    0.2126 * srgbChannelToLinear(r) +
    0.7152 * srgbChannelToLinear(g) +
    0.0722 * srgbChannelToLinear(b)
  );
}

export function contrastRatio(hexA: string, hexB: string): number {
  const luminanceA = relativeLuminance(hexA);
  const luminanceB = relativeLuminance(hexB);
  const lighter = Math.max(luminanceA, luminanceB);
  const darker = Math.min(luminanceA, luminanceB);
  return (lighter + 0.05) / (darker + 0.05);
}

export function meetsWcagAA(ratio: number, largeText = false): boolean {
  return ratio >= (largeText ? 3.0 : 4.5);
}

export function meetsWcagNonText(ratio: number): boolean {
  return ratio >= 3.0;
}

File: services/langgraph/agency/artifacts/color_contrast.py
"""WCAG 2.2 contrast-ratio math -- the real formula, not an approximation.

Relative luminance and contrast ratio as defined by WCAG 2.x/2.2:
sRGB channel -> linear-light value -> luminance-weighted sum, then
contrast = (lighter + 0.05) / (darker + 0.05).

These utilities are for cases where WCAG contrast actually applies -- body
text, UI components, graphical objects (SC 1.4.11). They are deliberately
NOT wired into ``LogoLockup`` validation: WCAG 2.2 SC 1.4.11 explicitly
exempts logotypes ("Text that is part of a logo or brand name has no
minimum contrast requirement"), so enforcing a contrast floor on a logo
lockup would be citing a compliance requirement that does not exist.
"""

from __future__ import annotations

import re

HEX_COLOR_RE = re.compile(r"^#[0-9A-Fa-f]{6}$")


def _srgb_channel_to_linear(channel_255: int) -> float:
    c = channel_255 / 255.0
    if c <= 0.03928:
        return c / 12.92
    return ((c + 0.055) / 1.055) ** 2.4


def relative_luminance(hex_color: str) -> float:
    """WCAG relative luminance of a 6-digit hex color, in [0, 1]."""
    if not HEX_COLOR_RE.match(hex_color):
        raise ValueError(f"'{hex_color}' is not a 6-digit hex color (e.g. '#1a2b4c')")
    value = hex_color.lstrip("#")
    r, g, b = (int(value[i : i + 2], 16) for i in (0, 2, 4))
    r_lin, g_lin, b_lin = (_srgb_channel_to_linear(channel) for channel in (r, g, b))
    return 0.2126 * r_lin + 0.7152 * g_lin + 0.0722 * b_lin


def contrast_ratio(hex_a: str, hex_b: str) -> float:
    """WCAG contrast ratio between two hex colors, in [1, 21]. Symmetric."""
    luminance_a = relative_luminance(hex_a)
    luminance_b = relative_luminance(hex_b)
    lighter, darker = max(luminance_a, luminance_b), min(luminance_a, luminance_b)
    return (lighter + 0.05) / (darker + 0.05)


def meets_wcag_aa(ratio: float, *, large_text: bool = False) -> bool:
    """SC 1.4.3: 4.5:1 for normal text, 3:1 for large text (>=18pt or >=14pt bold)."""
    return ratio >= (3.0 if large_text else 4.5)


def meets_wcag_non_text(ratio: float) -> bool:
    """SC 1.4.11: 3:1 for UI components and graphical objects (not logotypes)."""
    return ratio >= 3.0
